### DISCUSSION TITLE
Run github actions before merge

### DISCUSSION THREAD

**Discussion by jacobwojcik** [community/community#53922] [Actions] [answered]
### Select Topic Area

Question

### Body

Hi, I want to run an expensive job only before merge to `main` branch, so if the job fails the merge is canceled.
I know about 
```
on:
  push:
    branches:
      - master
  ```
  
  But it runs after merge and I want protect the `main` branch.

**Comment by multikitty** [comment]
Hi, @jacobwojcik 
```
on:
  pull_request:
    branches:
      - master
```
try pull_request instead of push
hope this will help you.

**Reply by jacobwojcik** [reply]
Does it run after every push in pr branch? Because I want prevent it

**Comment by multikitty** [comment]
github action runs when a pull request is opened on the  `main`  branch.  specifically, triggered when the  `pull_request`  event occurs on the main branch, and the workflow's  `on`  parameter filters the events to only run in that scenario. 
this means , the workflow will run whenever a pull request is opened on the  `main`  branch, and it will not run on every push to the pull request branch. 
howevr, you can modify the  `on`  parameter to specify different trigger events, such as  `pull_request_target`  or  `push` , depending on your specific requirements. 
by doing so, you can customize the workflow to run based on specific changes to the pull request branch.

Please have a try and check.

**Reply by HalDavis2** [reply]
is there a .net varify website in MSBuild

**Comment by jsoref** [comment, marked as answer]
Try merge queues:
https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue#triggering-merge-group-checks-with-github-actions

**Reply by MareoRaft** [reply]
Please note that merge queues are only available for:
* private repositories in an organization paying for GitHub Enterprise
* public repositories in any organization

So personal repositories don’t support that feature at all, nor private repositories in an org without GitHub Enterprise.

**Reply by jsoref** [reply]
Sure. Also, people really really shouldn't use personal repositories. Create an org, you get more features at no cost (including a better permission model as well as the ability to see which repositories are burning runners).

**Comment by ghost** [comment]
Not sure if [Merge queue](https://docs.github.com/en/repositories/configuring-branches-and-merges-in-your-repository/configuring-pull-request-merges/managing-a-merge-queue) are flexible enough or maybe I miss something.

1. We need to set a set of jobs names which should report their status
2. Jobs names for PR should be equal to Job names for Merge queue
3. Or we should run PR jobs and Merge queue jobs at both events - `merge_group` and `pull_request` which is unacceptable

In case we have a different set of jobs and their names, especially in matrix, a single way to do that is to use same matrix which is unacceptable (p.3 above).

@jacobwojcik, how did you solve that for your case?

**Reply by jsoref** [reply]
Can you explain why 3 is a problem?

Certainly while you're reviewing a PR, you need to know if the contents of the PR are valid.

Once you've established that the contents of the PR are independently valid, surely you'll want to know if the contents of the PR are still valid with any new things that are being included after the PR was tested but before it actually merges.

**Reply by ghost** [reply]
It is the same as in the [initial message](#discussion-5138538)
> I want to run an expensive job only before merge to main branch, so if the job fails the merge is canceled.

We want to run some additional checks only before the merge and not at the PR creation or every next commits to it.

**Reply by ghost** [reply]
So, a simple solution which will check the status based on the job name `PR check`, will work with the example above.

But it is very ambiguous and limited solution with the same job name in a different workflows. Merge queue jobs names should be a separate option with own setting.

`pr.yaml`
```yaml
name: PR

on:
  pull_request:

jobs:
  pr:
    name: PR check
    runs-on: ubuntu-latest
    steps:
      - name: PR check
        run: |
          echo sleep 10
```

`mq.yaml`
```yaml
name: MQ

on:
  merge_group:

jobs:
  mq:
    name: PR check
    runs-on: ubuntu-latest
    steps:
      - name: Merge queue check
        run: sleep 120
```

We are not alone and people actively discuss similar issues - [merge queue](https://github.com/orgs/community/discussions?discussions_q=merge+queue).

And main thread - [Pull Request Merge Queue Public Beta Feedback #46757](https://github.com/orgs/community/discussions/46757).

**Comment by setoelkahfi** [comment]
> ### Select Topic Area
> 
> Question
> 
> ### Body
> 
> Hi, I want to run an expensive job only before merge to `main` branch, so if the job fails the merge is canceled.
> 
> I know about 
> 
> ```
> 
> on:
> 
>   push:
> 
>     branches:
> 
>       - master
> 
>   ```
> 
>   
> 
>   But it runs after merge and I want protect the `main` branch.

Did you find the solution? We have the same problem, as in our development branch has release build configuration than the PR branch. I want to run a last check after two approvals in a PR to prevent running the release configuration on every PR created.

**Reply by ghost** [reply]
We have discussed two possible workarounds for merge queue - [Merge queue specific checks #103114](https://github.com/orgs/community/discussions/103114).

Maybe it may fits your needs.

**Reply by setoelkahfi** [reply]
Thanks! I'll take a look
